src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -905,6 +905,23 @@ private static void ThrowKeyNullException()
             throw new ArgumentNullException("key");
         }
 
+
+        // Allow nulls for reference types and Nullable<U>, but not for value types.
+        // Aggressively inline so the jit evaluates the if in place and either drops the call altogether
+        // Or just leaves null test and call to the Non-returning ThrowValueNullException
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void ThrowIfNullAndNullsAreIllegal(object value)
+        {
+            // Note that default(T) is not equal to null for value types except when T is Nullable<U>.
+            if (!(default(TValue) == null) && value == null)
```

Comment:
Why `!(default(TValue) == null)` instead of `default(TValue) != null`?

Reply:
I took this method from coreclr. Will change it.